src/main/sensors/gyro.c

```diff
@@ -516,64 +521,65 @@ bool gyroInit(void)
-        }
+    if (gyroDetectSensor(&gyroSensor1, gyroDeviceConfig(0))) {
+        gyroDetectionFlags |= DETECTED_GYRO_1;
     }
 
-#ifdef USE_MULTI_GYRO
-    if (gyroToUse == GYRO_CONFIG_USE_GYRO_2 || gyroToUse == GYRO_CONFIG_USE_GYRO_BOTH) {
-        if (gyroInitSensor(&gyroSensor2, gyroDeviceConfig(1))) {
-            gyroHasOverflowProtection =  gyroHasOverflowProtection && gyroSensor2.gyroDev.gyroHasOverflowProtection;
-            detectionFlags |= DETECTED_GYRO_2;
-        }
+#if defined(USE_MULTI_GYRO)
+    if (gyroDetectSensor(&gyroSensor2, gyroDeviceConfig(1))) {
+        gyroDetectionFlags |= DETECTED_GYRO_2;
     }
+#endif
 
-    if (gyroToUse == GYRO_CONFIG_USE_GYRO_BOTH && detectionFlags != (DETECTED_GYRO_1 | DETECTED_GYRO_2)) {
-        // at least one gyro is missing.
+    if (gyroDetectionFlags == NO_GYROS_DETECTED) {
+        return false;
+    }
 
-        if (detectionFlags & DETECTED_GYRO_1) {
+#if defined(USE_MULTI_GYRO)
+    if ((gyroToUse == GYRO_CONFIG_USE_GYRO_BOTH && !(gyroDetectionFlags & DETECTED_BOTH_GYROS))
+        || (gyroToUse == GYRO_CONFIG_USE_GYRO_1 && !(gyroDetectionFlags & DETECTED_GYRO_1))
+        || (gyroToUse == GYRO_CONFIG_USE_GYRO_2 && !(gyroDetectionFlags & DETECTED_GYRO_2))) {
+        if (gyroDetectionFlags & DETECTED_GYRO_1) {
             gyroToUse = GYRO_CONFIG_USE_GYRO_1;
-            gyroConfigMutable()->gyro_to_use = GYRO_CONFIG_USE_GYRO_1;
-            detectedSensors[SENSOR_INDEX_GYRO] = gyroSensor1.gyroDev.gyroHardware;
-            sensorsSet(SENSOR_GYRO);
-            return true;
-        }
-
-        if (detectionFlags & DETECTED_GYRO_2) {
+        } else {
             gyroToUse = GYRO_CONFIG_USE_GYRO_2;
-            gyroConfigMutable()->gyro_to_use = GYRO_CONFIG_USE_GYRO_2;
-            detectedSensors[SENSOR_INDEX_GYRO] = gyroSensor2.gyroDev.gyroHardware;
-            sensorsSet(SENSOR_GYRO);
-            return true;
         }
+
+        gyroConfigMutable()->gyro_to_use = gyroToUse;
     }
 
     // Only allow using both gyros simultaneously if they are the same hardware type.
-    // If the user selected "BOTH" and they are not the same type, then reset to using only the first gyro.
-    if (gyroToUse == GYRO_CONFIG_USE_GYRO_BOTH) {
-        if (gyroSensor1.gyroDev.gyroHardware != gyroSensor2.gyroDev.gyroHardware) {
-            gyroToUse = GYRO_CONFIG_USE_GYRO_1;
-            gyroConfigMutable()->gyro_to_use = GYRO_CONFIG_USE_GYRO_1;
-            detectedSensors[SENSOR_INDEX_GYRO] = gyroSensor1.gyroDev.gyroHardware;
-            sensorsSet(SENSOR_GYRO);
-        }
+    if ((gyroDetectionFlags & DETECTED_BOTH_GYROS) && gyroSensor1.gyroDev.gyroHardware != gyroSensor2.gyroDev.gyroHardware) {
+        gyroDetectionFlags |= DETECTED_DUAL_GYROS;
+    } else if (gyroToUse == GYRO_CONFIG_USE_GYRO_BOTH) {
+        // If the user selected "BOTH" and they are not the same type, then reset to using only the first gyro.
+        gyroToUse = GYRO_CONFIG_USE_GYRO_1;
```

Comment:
How about the edge case where only gyro 2 was detected and the user had set to "BOTH"?

Reply:
Discrepancies in selected vs. detected gyros are checked and rectified in https://github.com/betaflight/betaflight/pull/7914/files#diff-59167518290ca93ea0a3471df4e48bbbR543, so if there aren't two gyros when we come here, `gyroToUse` won't be `GYRO_CONFIG_USE_GYRO_BOTH` any more.